### PR TITLE
Make permission queries optional

### PR DESCRIPTION
**Pull request by MrCordeiro**
PR related to issue #99: extra queries hit the database needlessly if the moderation feature is not being used.

### CODE REVIEW

**Comment by MrCordeiro** [comment]
I decided to add a second setting instead of using the number of flags, as I find both features are potentially independent: moderation could exist even if the user doesn't want to use the flags.

**Comment by abhiabhi94** [comment]
Hey @MrCordeiro, thanks for the contribution.

From the way, the comment application is written as of now, I think this will require change at a few more places.

I will point out the specific instances in the codebase, when I have some more time today.

**Comment by abhiabhi94** [comment]
> I find both features are potentially independent: moderation could exist even if the user doesn't want to use the flags.

In a situation when flagging is enabled and moderation isn't,  I don't understand the purpose of flagging then.


For all to be on the same page, If I understand correctly, by moderation you mean permissions and groups? Correct me if I didn't get you completely.

If that is in the case, then checking this be sufficient:
```python
def is_comment_admin:
    if settings.COMMENT_FLAG_ALLOWED:
        return user.groups.filter(name="comment_admin").exists() or (
            user.has_perm("comment.delete_flagged_comment")
            and user.has_perm("comment.delete_comment")
        )
    return False
```

**Comment by MrCordeiro** [comment]
Sounds good.
You told be there are other instances of the code that would require changing. Have you had the time to list them?

File: docs/source/settings.rst [hunk 1]
```diff
@@ -87,6 +87,10 @@ COMMENT_ALLOW_ANONYMOUS
 
 Should the anonymous commenting featured be allowed? Defaults to ``False``.
 
+COMMENT_ALLOW_MODERATION
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Should there be user roles for moderation and admin? If not, set this to `False` in order not to query the database unnecessarily. Defaults to ``True``.
```
Comment (abhiabhi94) [inline docs/source/settings.rst:93]: Please put double back ticks around `False`.
```suggestion
Should there be user roles for moderation and admin? If not, set this to ``False`` in order not to query the database unnecessarily. Defaults to ``True``.
```

File: comment/conf/defaults.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 COMMENT_USE_GRAVATAR = False
 
 COMMENT_FLAGS_ALLOWED = 0
+COMMENT_ALLOW_MODERATION = True
```
Comment (abhiabhi94) [inline comment/conf/defaults.py:11]: Maybe you can remove this now that an extra setting is no longer required.

File: comment/utils.py [hunk 3]
```diff
@@ -79,12 +79,20 @@ def has_valid_profile():
 
 
 def is_comment_admin(user):
-    return user.groups.filter(name='comment_admin').exists() or \
-           (user.has_perm('comment.delete_flagged_comment') and user.has_perm('comment.delete_comment'))
+    if settings.COMMENT_ALLOW_MODERATION:
```
Comment (abhiabhi94) [inline comment/utils.py:82]:
```suggestion
    if settings.COMMENT_FLAGS_ALLOWED`:
```

File: comment/utils.py [hunk 4]
```diff
@@ -79,12 +79,20 @@ def has_valid_profile():
...
+        return user.groups.filter(name="comment_admin").exists() or (
+            user.has_perm("comment.delete_flagged_comment")
+            and user.has_perm("comment.delete_comment")
+        )
+    return False
 
 
 def is_comment_moderator(user):
-    return user.groups.filter(name='comment_moderator').exists() or user.has_perm('comment.delete_flagged_comment')
+    if settings.COMMENT_ALLOW_MODERATION:
```
Comment (abhiabhi94) [inline comment/utils.py:91]:
```suggestion
    if settings.COMMENT_FLAGS_ALLOWED`:
```

**Review by abhiabhi94 — CHANGES_REQUESTED** [review]
In addition to the already made changes, inside the [`post_migrate signals module`](https://github.com/Radi85/Comment/blob/master/comment/signals/post_migrate.py#L8), the `permission_groups` need not be created nor is there any need to running `adjusting_flagged_comment`. If the developer chooses to have moderation at a later stage, they can toggle the setting and run `./migrate comment` again.

The documentation regarding flagging now, maybe needs to be moved to a separate file. It would state the steps to take when enabling flagging. Although you can safely ignore it, I will add it after this is merged.

File: comment/utils.py (hunk 3 above)
Comment (MrCordeiro) [reply comment/utils.py:82]: I think there's an extra " ` " at the end there, no?

**Review by MrCordeiro — COMMENTED** [review]
PR updated according to the comments. 
I also modified the `post_migrate` signal with a similar condition check. However, was unable to write an unit test for it. I have never written a migration test 😬, so I would appreciate any input/help 🙏

**Comment by abhiabhi94** [comment]
> Regarding the migrations test

I assume it would be a bit difficult to test this change because the current `BaseMigrationTest` class works well when we have to go in the migration history from one state to the last state(suppose from 5 to 10). As of now, it takes the name of the migration module to go to. In this case, we'll have to revert to the zeroth state because all permissions and stuff is created whenever the migration is run(we migrate before any test is run, see [`tox.ini`](https://github.com/Radi85/Comment/blob/master/tox.ini#L42) for reference). 

One of the alternative that I think of is adding a separate module for testing. As of now, the `example` application and testing both use the same module. The `testing` module will have flagging disabled by default. You can safely ignore this, I will try to see what best we can do to cover this.


To allow your failing test to pass, you will have to change the function `test_change_state_when_flagging_is_disabled`. This test wasn't actually doing what the test name said :(

Also, you'll have to add a test to cover the new changes.

```python
    # fix for the failing test
    @patch.object(settings, 'COMMENT_FLAGS_ALLOWED', False)
    def test_change_state_when_flagging_is_disabled(self):
        response = self.client.post(self.get_url(), data=self.data)
        self.assertEqual(response.status_code, 403)

    # new test
    @patch.object(settings, 'COMMENT_FLAGS_ALLOWED', 1)
    def test_change_state_when_comment_is_not_flagged(self):
        comment = self.comment_2
        self.assertFalse(comment.is_flagged)
        response = self.client.post(self.get_url(comment.id), data=self.data)
        self.assertEqual(response.status_code, 400)
``` 

There is a bit of cleanup required with these `mixins` and `models`, that I forgot. I will do it after this is merged.

**Comment by abhiabhi94** [comment]
Also, when you are done making the changes, can you please squash your changes to one commit. To keep the git history clean, it would be nice to have this change inside one commit.

**Comment by MrCordeiro** [comment]
But isn't the squash applied as an option when merging the PR?

**Review by MrCordeiro — COMMENTED** [review]
Updated the tests results

**Comment by abhiabhi94** [comment]
> But isn't the squash applied as an option when merging the PR?

Yes, but it will also include all the messages from all of your commits. I will do it if you have an issue with it.

Thanks for the contribution.

File: comment/tests/test_api.py [hunk 5]
```diff
@@ -504,6 +504,13 @@ def test_change_state_when_flagging_is_disabled(self):
         comment = self.comment_2
         self.assertFalse(comment.is_flagged)
         response = self.client.post(self.get_url(), data=self.data)
+        self.assertEqual(response.status_code, 403)
+        
+    @patch.object(settings, 'COMMENT_FLAGS_ALLOWED', 1)
+    def test_change_state_when_comment_is_not_flagged(self):
+        comment = self.comment_2
+        self.assertFalse(comment.is_flagged)
```
Comment (abhiabhi94) [inline comment/tests/test_api.py:512]: These two lines are required inside this function as now. You may want to remove them,

**Review by MrCordeiro — COMMENTED** [review]
Squashed commits

**Comment by abhiabhi94** [comment]
@MrCordeiro can you please fix the `flake8` errors.

**Review by abhiabhi94 — APPROVED** [review]

**Comment by abhiabhi94** [comment]
Thanks !

**Comment by MrCordeiro** [comment]
Sorry, it took so long. I had totally forgotten about this PR until your last comment.
________________________________

**Review by Radi85 — APPROVED** [review]

**Pull request merged**